Native assets support for MacOS and iOS

dcharkes opened the pull request:
Support for FFI calls with `@Native external` functions through Native assets on MacOS and iOS. This enables bundling native code without any build-system boilerplate code.

For more info see:

* https://github.com/flutter/flutter/issues/129757

### Implementation details for MacOS and iOS.

Dylibs are bundled by (1) making them fat binaries if multiple architectures are targeted, (2) code signing these, and (3) copying them to the frameworks folder. These steps are done manual rather than via CocoaPods. CocoaPods would have done the same steps, but (a) needs the dylibs to be there before the `xcodebuild` invocation (we could trick it, by having a minimal dylib in the place and replace it during the build process, that works), and (b) can't deal with having no dylibs to be bundled (we'd have to bundle a dummy dylib or include some dummy C code in the build file).

The dylibs are build as a new target inside flutter assemble, as that is the moment we know what build-mode and architecture to target.

The mapping from asset id to dylib-path is passed in to every kernel compilation path. The interesting case is hot-restart where the initial kernel file is compiled by the "inner" flutter assemble, while after hot restart the "outer" flutter run compiled kernel file is pushed to the device. Both kernel files need to contain the mapping. The "inner" flutter assemble gets its mapping from the NativeAssets target which builds the native assets. The "outer" flutter run get its mapping from a dry-run invocation. Since this hot restart can be used for multiple target devices (`flutter run -d all`) it contains the mapping for all known targets.

### Example vs template

The PR includes a new template that uses the new native assets in a package and has an app importing that. Separate discussion in: https://github.com/flutter/flutter/issues/131209.

### Tests

This PR adds new tests to cover the various use cases.

* dev/devicelab/bin/tasks/native_assets_ios.dart
  * Runs an example app with native assets in all build modes, doing hot reload and hot restart in debug mode.
* dev/devicelab/bin/tasks/native_assets_ios_simulator.dart
  * Runs an example app with native assets, doing hot reload and hot restart.
* packages/flutter_tools/test/integration.shard/native_assets_test.dart
  * Runs (incl hot reload/hot restart), builds, builds frameworks for iOS, MacOS and flutter-tester.
* packages/flutter_tools/test/general.shard/build_system/targets/native_assets_test.dart
  * Unit tests the new Target in the backend.
* packages/flutter_tools/test/general.shard/ios/native_assets_test.dart
* packages/flutter_tools/test/general.shard/macos/native_assets_test.dart
  * Unit tests the native assets being packaged on a iOS/MacOS build.

It also extends various existing tests:

* dev/devicelab/bin/tasks/module_test_ios.dart
   * Exercises the add2app scenario.
* packages/flutter_tools/test/general.shard/features_test.dart
   * Unit test the new feature flag.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

dcharkes committed:
Native assets support for MacOS and iOS

dcharkes committed:
Merge branch 'master' into native-assets-macos-ios

dcharkes committed:
Updated pubspecs

dcharkes committed:
Fix analyzer warnings

dcharkes committed:
fix comment

dcharkes committed:
Fix printing native assets as null

dcharkes committed:
Don't pass empty native asset paths on

dcharkes committed:
Check native assets folder exists before rsync

dcharkes committed:
fix whitespace

dcharkes committed:
Move shared test infra to separate file

dcharkes committed:
Use mocked FileSystem in BuildIOS commands

dcharkes committed:
fix test mock file system

dcharkes committed:
Fix another test mock file system

dcharkes committed:
Skip test on Linux and Windows

dcharkes committed:
fix analysis

dcharkes committed:
Don't try to build native assets if there is no package config + fix some tests

dcharkes committed:
remove unused imports

dcharkes committed:
fix test

dcharkes committed:
print stdout and stderr in test on failure

dcharkes committed:
fix test

dcharkes committed:
Merge commit 'f212b1012de72d5a8e041f768e0a9026b919060d' into native-assets-macos-ios

dcharkes committed:
Run `flutter update-packages --force-upgrade`.

vashworth reviewed (changes requested):
The logic seems sound to me expect for a few points I mention in my comments.

Couple other things, though: 

Question: Is the template meant to be used by users or just added added for the sake of testing?

### Formatting
I noticed a lot of formatting where you line break after the assignment operator. I didn’t leave comments because there were a bunch and I didn’t want to clutter from the important comments. Flutter style guideline is to [Prefer avoiding line breaks after assignment operators](https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#prefer-a-maximum-line-length-of-80-characters).

Similarly, [we use trailing commas for arguments and parameters and give each their own line](https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#use-a-trailing-comma-for-arguments-parameters-and-list-items-but-only-if-they-each-have-their-own-line).

### Add to App
Seems to be this will only work with [Option B for Add to App](https://docs.flutter.dev/add-to-app/ios/project-setup#option-b---embed-frameworks-in-xcode). Is this going to be a known limitation that you can only use native assets with Add to App if you do Option B?

### Tests
I didn’t see any macOS devicelab tests. Any reason why not? If not, add them?

Also, the following should probably be tested:

Test changes made to create.dart in packages/flutter_tools/test/commands.shard/permeable/create_test.dart

Test changes made to compile.dart in packages/flutter_tools/test/general.shard/compile_batch_test.dart

Test changes made to features.dart in packages/flutter_tools/test/general.shard/features_test.dart

Add tests to ensure dry run happens in the following tests:
packages/flutter_tools/test/general.shard/ios/mac_test.dart
packages/flutter_tools/test/commands.shard/hermetic/build_macos_test.dart
packages/flutter_tools/test/general.shard/hot_test.dart

Also, add unit tests for 
packages/flutter_tools/lib/src/build_system/targets/native_assets.dart
packages/flutter_tools/lib/src/ios/native_assets.dart
packages/flutter_tools/lib/src/macos/native_assets.dart

vashworth commented on dev/devicelab/bin/tasks/native_assets_ios.dart:
```diff
@@ -0,0 +1,14 @@
+import 'package:flutter_devicelab/framework/framework.dart';
+import 'package:flutter_devicelab/tasks/native_assets_test.dart';
+
+Future<void> main() async {
```
Also, if you want it to run in presubmit, you'll need to file an infra ticket (https://github.com/flutter/flutter/tree/master/dev/devicelab#adding-tests-to-presubmit)

vashworth commented on dev/devicelab/bin/tasks/native_assets_ios_simulator.dart:
```diff
@@ -0,0 +1,31 @@
+import 'package:flutter_devicelab/framework/task_result.dart';
+import 'package:flutter_devicelab/tasks/native_assets_test.dart';
+
+Future<void> main() async {
```
Also, if you want it to run in presubmit, you'll need to file an infra ticket (https://github.com/flutter/flutter/tree/master/dev/devicelab#adding-tests-to-presubmit)

dcharkes commented:
Thanks @vashworth !

Some high level replies to your questions below.

> Question: Is the template meant to be used by users or just added added for the sake of testing?

I'd like either a sample project or a template.

A template is more discoverable than a sample project. I worry that we will push users towards using the old (and when native assets are no longer experimental) FFI plugin template if we don't have a new template.

(A sample project could be in https://github.com/dart-lang/native/tree/main/pkgs/native_assets_cli/example, however having the Flutter app also there would mean we cannot unit test it on the CI of that repo.)

Maybe we can somehow get all the benefits of a template without all the machinery? Suggestions?

> Seems to be this will only work with [Option B for Add to App](https://docs.flutter.dev/add-to-app/ios/project-setup#option-b---embed-frameworks-in-xcode). Is this going to be a known limitation that you can only use native assets with Add to App if you do Option B?

I wasn't aware of the other options.

Maybe we can make the `podhelper.rb` also do copying. I had that in an earlier prototype I believe. (Though I'm not so fluent in Ruby so that was crashing when there were no native assets. And it requires us to pass some information somewhere.)

> I didn’t see any macOS devicelab tests. Any reason why not? If not, add them?

All the MacOS tests are in test/integration.shard/native_assets_test.dart

Maybe that is the wrong place for tests that run a MacOS application? Should they be in a different place?

Also, currently that test also runs `flutter build ios`, but it [fails on the CI](https://logs.chromium.org/logs/flutter/buildbucket/cr-buildbucket/8774707149105331809/+/u/run_test.dart_for_tool_integration_tests_shard_and_subshard_3_4/test_stdout):

```
02:57 +26 -1: test/integration.shard/native_assets_test.dart: flutter build ios with native assets release [E]                                                                                         
  Exception: flutter build failed: 1
  Error (Xcode): No signing certificate "iOS Development" found: No "iOS Development" signing certificate matching team ID "S8QB4VV633" with a private key was found.
```

Maybe this is also a result of running a test on the wrong shard/bot? Should building ios apps also happen in the iOS device lab instead of on an ordinary MacOS machine?

dcharkes committed:
Merge branch 'main' into native-assets-macos-ios

dcharkes committed:
Run `flutter update-packages --force-upgrade`

dcharkes committed:
Add ios devicelab tests to ci

dcharkes commented on dev/devicelab/lib/tasks/native_assets_test.dart:
```diff
@@ -0,0 +1,223 @@
+  await process.exitCode;
+}
+
+Future<void> _checkAppRunning(bool shouldBeRunning) async {
```
Removed, we should just be waiting on the process to exit.

dcharkes commented on packages/flutter_tools/lib/src/ios/native_assets.dart:
```diff
@@ -0,0 +1,152 @@
+Asset _targetLocationiOS(Asset asset) {
+  final AssetPath path = asset.path;
+  switch (path) {
+    case AssetSystemPath _:
```
Yes. I learned about it through a lint. A lint started complaining on the old `.runtimeType` and `AssetSystemPath`.

dcharkes committed:
address small comments

zanderso commented on packages/flutter_tools/test/src/common.dart:
```diff
@@ -184,6 +184,9 @@ void test(String description, FutureOr<void> Function() body, {
     // definitely be enough.
+    // Specifying a timeout here as well enables running test files
+    // standalone without `flutter test`.
+    timeout: const Timeout(Duration(minutes: 15)),
```
DBC

It would be confusing to change the number in `dart_test.yaml`, and for that change to have no effect.

dcharkes replied:
We can leave cross-comments here and there.

If I remove this line running a test with `dart path/to/my_test.dart` times out. Running `flutter test` instead is slower due to having to precompile some package:test machinery.

I'm happy to revert this before I land this PR if you'd really don't want to support the `dart path/to/my_test.dart`.

zanderso replied:
> If I remove this line running a test with `dart path/to/my_test.dart` times out.

What's the root cause of that? Is the default timeout too short?

> support the `dart path/to/my_test.dart`.

If it isn't tested on CI, it will likely break some other way even if this line is added. If this is something that you'd like to keep working, you'd need to add a test to CI for it. Given the size of this change, that might be better to do as a separate PR.

dcharkes replied:
The root cause is individual tests having a timeout duration of 30 seconds by default.

```
$ dart /Users/dacoharkes/flt/engine/flutter/packages/flutter_tools/test/integration.shard/native_assets_test.dart
00:00 +0: (setUpAll)

00:00 +0: flutter run hot reload and hot restart with native assets flutter-tester debug

00:16 +1: flutter run hot reload and hot restart with native assets macos debug

00:41 +2: flutter run with native assets macos profile

01:11 +2 -1: flutter run with native assets macos profile [E]

  TimeoutException after 0:00:30.000000: Test timed out after 30 seconds. See https://pub.dev/packages/test#timeouts

  package:test_api/src/backend/invoker.dart 338:28  Invoker._handleError.<fn>
  dart:async/zone.dart 1391:47                      _rootRun
  dart:async/zone.dart 1301:19                      _CustomZone.run
  package:test_api/src/backend/invoker.dart 336:10  Invoker._handleError
  package:test_api/src/backend/invoker.dart 291:9   Invoker.heartbeat.<fn>.<fn>
  dart:async/zone.dart 1399:13                      _rootRun
  dart:async/zone.dart 1301:19                      _CustomZone.run
  package:test_api/src/backend/invoker.dart 290:38  Invoker.heartbeat.<fn>
  dart:async-patch/timer_patch.dart 18:15           Timer._createTimer.<fn>
  dart:isolate-patch/timer_impl.dart 398:19         _Timer._runTimers
  dart:isolate-patch/timer_impl.dart 429:5          _Timer._handleMessage
  dart:isolate-patch/isolate_patch.dart 184:12      _RawReceivePort._handleMessage
  

01:11 +2 -1: flutter run with native assets macos release

[...]

$ flutter test /Users/dacoharkes/flt/engine/flutter/packages/flutter_tools/test/integration.shard/native_assets_test.dart
[ ... works fine ]
```

> > support the `dart path/to/my_test.dart`.
> 
> If it isn't tested on CI, it will likely break some other way even if this line is added. If this is something that you'd like to keep working, you'd need to add a test to CI for it. Given the size of this change, that might be better to do as a separate PR.

That might be a bit overkill, testing that we can run a test as a standalone dart file instead of with `flutter test`. I'll revert the change.

dcharkes committed:
Try fixing CI

zanderso reviewed (commented):
The scope of the feature doesn't seem large enough to support its own template. I'd prefer adding something to https://github.com/flutter/flutter/tree/master/examples, which is where the other platform interop examples live.

dcharkes committed:
more formatting

dcharkes committed:
replace `assert(` with `throwToolExit(`

dcharkes committed:
formatting

dcharkes committed:
Don't rsync if no native assets dir

dcharkes committed:
test owners

dcharkes committed:
Use `--no-codesign` for non-devicelab test

vashworth commented:
@dcharkes I looked over the App to App solutions again, and I think it should work with your design actually. [Option A](https://docs.flutter.dev/add-to-app/ios/project-setup#option-a---embed-with-cocoapods-and-the-flutter-sdk) adds script phases which call xcode_backend.sh, like a normal flutter project, so your code should be fine.

https://github.com/flutter/flutter/blob/master/packages/flutter_tools/templates/module/ios/library/Flutter.tmpl/podhelper.rb.tmpl#L92-L100

And looking closer at [Option C](https://docs.flutter.dev/add-to-app/ios/project-setup#option-c---embed-application-and-plugin-frameworks-in-xcode-and-flutter-framework-with-cocoapods), plugins still need to be embedded manually so the same would be true for native assets, so that also should be fine.

Therefore, I believe all three options should be compatible. 

Also, here's an integration tests used in the tool for add to app (Looks like it uses Option A): https://github.com/flutter/flutter/tree/master/dev/integration_tests/ios_add2app_life_cycle 
https://github.com/flutter/flutter/blob/5554b0eeb3780d3fac90370d75dc57893733ae71/dev/bots/test.dart#L762-L773

Looks like Option B is only tested by ensuring the frameworks are created and exist (https://github.com/flutter/flutter/blob/master/dev/devicelab/bin/tasks/build_ios_framework_module_test.dart)

dcharkes committed:
Remove mock checking hack

dcharkes committed:
Follow multi-step ci.yaml plan

dcharkes committed:
Merge template create logic

dcharkes committed:
Fix template readme

dcharkes committed:
Don't generate OS-files for ffi packages

dcharkes committed:
Use configurable build directory

dcharkes committed:
Fix FFI plugin create

dcharkes committed:
Support `flutter run -d all`

dcharkes committed:
Fix formatting

dcharkes committed:
Set required environment variable for backend test

zanderso commented on packages/flutter_tools/templates/package_ffi/build.dart.tmpl:
```diff
@@ -0,0 +1,28 @@
+// Copyright (c) 2023, the Dart project authors.  Please see the AUTHORS file
```
Is this code copied from the Dart repo? Otherwise this should have the flutter/flutter header, I think.

dcharkes committed:
Remove copyright from template

dcharkes committed:
Remove timeout override

dcharkes commented:
> Therefore, I believe all three options should be compatible.

Great!

> Also, here's an integration tests used in the tool for add to app (Looks like it uses Option A): https://github.com/flutter/flutter/tree/master/dev/integration_tests/ios_add2app_life_cycle

I'm not sure that this test is actually run, as it is broken for me locally. I've filed https://github.com/flutter/flutter/issues/131261 to follow up.

> Looks like Option B is only tested by ensuring the frameworks are created and exist (https://github.com/flutter/flutter/blob/master/dev/devicelab/bin/tasks/build_ios_framework_module_test.dart)

The PR has a test similar to that in test/integration.shard/native_assets_test.dart, see `expectDylibIsBundledWithFrameworks`.

dcharkes committed:
Add create tests.

dcharkes committed:
compile test

dcharkes committed:
features test

dcharkes committed:
Fix formatting

dcharkes committed:
fix formatting

dcharkes committed:
Merge branch 'main' into native-assets-macos-ios

dcharkes committed:
native asset target unit tests

dcharkes committed:
native asset dry run and build unit tests

dcharkes commented:
> I believe approach (1) might be better, because we'd actually be testing that flutter_tools invokes the dart path/to/some/build.dart of the various packages that have native assets that need to be build. With approach (2) we'd just be returning some boolean or a path to a native_assets.yaml which would not be catching many errors.

With approach (1) we'd have to mock a variety of process calls and create a bunch of files on these process calls (https://github.com/dart-lang/native/issues/90#issuecomment-1651352057). That would make these tests rather involved,  mimicking pub and other tools.

I'm not entirely sure where to put the mock boundary with approach (2). We could wrap both `NativeAssetsBuilder` and `PackageConfig` and make mocks for those. But that would also mean these would have to be passed around everywhere or be dependency injected through `Environment`. Passing around seems like a bad idea. But I also don't see any other examples of dependencies of flutter_tools being dependency injected. Do we have an example of mocking a dependency somewhere in flutter_tools? @stuartmorgan @vashworth.

Without mocking, the unit testing as requested is rather limited:

> Add tests to ensure dry run happens in the following tests:
> * packages/flutter_tools/test/general.shard/ios/mac_test.dart
> * packages/flutter_tools/test/commands.shard/hermetic/build_macos_test.dart
> * packages/flutter_tools/test/general.shard/hot_test.dart

These tests already cover the paths where there are no native assets, and all the extra code (a path to a yaml, a xcode property etc.) is null in that case.

> Also, add unit tests for
>
> * packages/flutter_tools/lib/src/build_system/targets/native_assets.dart 
> * packages/flutter_tools/lib/src/ios/native_assets.dart
> * packages/flutter_tools/lib/src/macos/native_assets.dart

I've added unit tests for these, but they can only unit test the behavior that does not actually invoke the `package:native_assets_builder` dependency, so the unit tests are rather limited.  (The integration tests for MacOS and iOS should cover all this behavior.)

For the add2app test alternative I've run into another existing issue. https://github.com/flutter/flutter/issues/112033

The three remaining things before this PR can be merged:

1. Unit tests - Blocked by how to mock `package:native_assets_builder`. I need some guidance/example of how to do this in flutter_tools.
2. Add2app integration test actually exercising the bundled libraries - blocked by https://github.com/flutter/flutter/issues/112033 and https://github.com/flutter/flutter/issues/80807. I'm not sure I know how to make progress here.
3. Code formatting - I'd rather do this as the final step.

dcharkes committed:
skip justification comments

dcharkes committed:
remove skipped tests

vashworth commented:
> I'm not entirely sure where to put the mock boundary with approach (2). We could wrap both `NativeAssetsBuilder` and `PackageConfig` and make mocks for those. But that would also mean these would have to be passed around everywhere or be dependency injected through `Environment`. Passing around seems like a bad idea. But I also don't see any other examples of dependencies of flutter_tools being dependency injected. Do we have an example of mocking a dependency somewhere in flutter_tools?

I'm not sure of any examples of mocking a dependency. @christopherfujino Do you know of any?

dcharkes committed:
Make module_test_ios work locally

christopherfujino commented:
> > Add tests to ensure dry run happens in the following tests:
> > packages/flutter_tools/test/general.shard/ios/mac_test.dart
> > packages/flutter_tools/test/commands.shard/hermetic/build_macos_test.dart
> > packages/flutter_tools/test/general.shard/hot_test.dart
> 
> What does dry-run mean in this context @vashworth? Tests which run with the memory file system and mock process run?
> 
> I believe I would need to either (1) address [dart-lang/native#90](https://github.com/dart-lang/native/issues/90), or (2) make the code in packages/flutter_tools/lib/src/macos/native_assets.dart and packages/flutter_tools/lib/src/ios/native_assets.dart sit inside a class, and create mocks for that class and add that class as a field to `class BuildCommand` and a bunch of other places.

Could we create a fake `NativeAssetsBuildRunner` that we pass to the command?

Also, I think you should create a wrapper class anyway that you pass in the logger, processManager, and fileSystem, so that for tests we can pass test instances. We're migrating the codebase off of using any more `globals.*` getters, and introducing new invocations is not allowed.

dcharkes commented:
> > > Add tests to ensure dry run happens in the following tests:
> > > packages/flutter_tools/test/general.shard/ios/mac_test.dart
> > > packages/flutter_tools/test/commands.shard/hermetic/build_macos_test.dart
> > > packages/flutter_tools/test/general.shard/hot_test.dart
> > 
> > 
> > What does dry-run mean in this context @vashworth? Tests which run with the memory file system and mock process run?
> > I believe I would need to either (1) address [dart-lang/native#90](https://github.com/dart-lang/native/issues/90), or (2) make the code in packages/flutter_tools/lib/src/macos/native_assets.dart and packages/flutter_tools/lib/src/ios/native_assets.dart sit inside a class, and create mocks for that class and add that class as a field to `class BuildCommand` and a bunch of other places.
> 
> Could we create a fake `NativeAssetsBuildRunner` that we pass to the command?
> 
> Also, I think you should create a wrapper class anyway that you pass in the logger, processManager, and fileSystem, so that for tests we can pass test instances. We're migrating the codebase off of using any more `globals.*` getters, and introducing new invocations is not allowed.

Okay, I'll give this a spin tomorrow. It does mean we have to add a `NativeAssetsBuildRunner` to every class/function (e.g. all the `Command`s for example) from `main` and where we pass it in tests and we want to exercise the mock. Without some kind of dependency injection (a la globals) that might be a lot of threading.

Thinking of it, the `FakeResidentCompiler` is a mock of an external dependency. This one is passed as a field of `FlutterDevice`.

dcharkes committed:
Introduce a wrapper class for `NativeAssetsBuildRunner`

dcharkes committed:
Unit test xcconfig native assets

dcharkes committed:
Add `NativeAssetsBuildRunner` to `Command`s and add tests

dcharkes committed:
Run hot native assets test

dcharkes committed:
Fix analysis

dcharkes committed:
implement `packagesWithNativeAssets`

dcharkes committed:
fix formatting and imports

dcharkes committed:
Mock xcrun clang --version process calls in tests

dcharkes committed:
Print file contents in test to see why Windows fails

dcharkes committed:
Try fixing test on Window

dcharkes committed:
Merge branch 'main' into native-assets-macos-ios

dcharkes committed:
Fix another path in tests on Windows

dcharkes committed:
Run hot test

dcharkes commented:
> Could we create a fake `NativeAssetsBuildRunner` that we pass to the command?
> 
> Also, I think you should create a wrapper class anyway that you pass in the logger, processManager, and fileSystem, so that for tests we can pass test instances. We're migrating the codebase off of using any more `globals.*` getters, and introducing new invocations is not allowed.

Done. This is a bit verbose though adding fields and arguments in methods/constructor calls. @christopherfujino is this the preferred way of doing this?

This does not enable testing the `NativeAssets` `Target`. The targets all have const constructors and all use the dependency injection through globals. The documentation on `Target` reads:

```dart
/// ## Unit testing
///
/// Most targets will invoke an external binary which makes unit testing
/// trickier. It is recommend that for unit testing that a Fake is used and
/// provided via the dependency injection system. a [Testbed] may be used to
/// set up the environment before the test is run. Unit tests should fully
/// exercise the rule, ensuring that the existing input and output verification
/// logic can run, as well as verifying it correctly handles provided defines
/// and meets any additional contracts present in the target.
```

If I'm not mistaken the only way to use those overrides from a `context` are through `globals` such as `ProcessManager get processManager => context.get<ProcessManager>() ?? _kLocalProcessManager;`. Any suggestions @christopherfujino on how to add unit tests for a target that needs to mock its dependencies?

> packages/flutter_tools/test/general.shard/ios/mac_test.dart

I believe the tests should be in test/commands.shard/hermetic/build_ios_test.dart instead, no invocations of `buildXcodeProject` are tested in packages/flutter_tools/test/general.shard/ios/mac_test.dart. I've added a test in test/commands.shard/hermetic/build_ios_test.dart.

> packages/flutter_tools/test/commands.shard/hermetic/build_macos_test.dart

Done.

> packages/flutter_tools/test/general.shard/hot_test.dart

I could only figure out how to do a hot restart in here. `run` isn't implemented in the fake

I also tried add a test to test/general.shard/resident_runner_test.dart (another place where `HotRunner` is constructed and run). I could only figure out how to `run`. For `restart` and `reload` I can't figure out what the `fakeVmServiceHost` expects. Maybe this test has the wrong fakes for just testing native assets. @vashworth what exactly would you like to seen unit tested for the hot runner? Is there already a similar test?

> Also, add unit tests for
> 
> packages/flutter_tools/lib/src/build_system/targets/native_assets.dart

See discussion above about dependency injection.

> packages/flutter_tools/lib/src/ios/native_assets.dart
> packages/flutter_tools/lib/src/macos/native_assets.dart

Done.

dcharkes committed:
Stop reusing dep object for now

dcharkes committed:
Merge branch 'main' into native-assets-macos-ios

dcharkes committed:
Add native assets to module_test_ios

dcharkes committed:
Enable native assets in module_test_ios

dcharkes commented:
> @dcharkes If you want to try and add this, you can go for it. Alternatively, it looks like `ios_host_app` is also an Add to App integration test (used in `module_test_ios.dart`). Perhaps use that?

_Originally posted in https://github.com/flutter/flutter/issues/131261#issuecomment-1650058595_

It seems that we lose the `NATIVE_ASSETS` env var setting the path to `native_assets.yaml` for the kernel compilation in `Build iOS Objective-C host app`. This is seen by the `flutter assemble` call having `-dNativeAssets= ` and no `NATIVE_ASSETS = ...` in the preceding lines. It looks like there is no xcconfig because the  `PACKAGE_CONFIG = ` in the log lines of these non-flutter originated invocations is missing (which is ones of the other defines in the xcconfig file). Is it a constraint for add2app that it should work without our generated xcconfig? @vashworth

> @dcharkes I looked over the App to App solutions again, and I think it should work with your design actually. [Option A](https://docs.flutter.dev/add-to-app/ios/project-setup#option-a---embed-with-cocoapods-and-the-flutter-sdk) adds script phases which call xcode_backend.sh, like a normal flutter project, so your code should be fine.

It doesn't seem entirely true. I believe there is no xcconfig which would set the right environment variables.

If we can't pass information through `xcconfig`, we have some options:

1. Do a dry-run invocation inside `KernelSnapshot` to produce a native_assets.yaml there. The downdside of this is that a `flutter run` invocation will now do a dry run both in the "outer" run and "inner" assemble invocation. Also it does a non-dry-run in the "inner" invocation.
   1.b. In addition keep passing native_assets.yaml in the xcconfig, and skip running in the `KernelSnapshot` if we already have one. (This requires us to pass an empty native_assets.yaml in projects without any native assets, instead of no native_assets.yaml that we do now.)
2. Make the `KernelSnapshot` target depend on the `NativeAssets` target and use the non-dry-run native_assets.yaml. The downside of this is that now `KernelSnapshot` caching doesn't work anymore. `NativeAssets` targets don't have caching.

dcharkes committed:
Fix bundling in add2app scenario

dcharkes committed:
Merge branch 'main' into native-assets-macos-ios

dcharkes committed:
Explore passing `native_assets.yaml` between targets

dcharkes committed:
Fix analysis issues

dcharkes committed:
Try using dependencies for native_assets_builder

vashworth commented:
> I also tried add a test to test/general.shard/resident_runner_test.dart (another place where HotRunner is constructed and run). I could only figure out how to run. For restart and reload I can't figure out what the fakeVmServiceHost expects. Maybe this test has the wrong fakes for just testing native assets. @vashworth what exactly would you like to seen unit tested for the hot runner? Is there already a similar test?

Yeah, looks like resident_runner_test.dart is the right place. I was thinking test to ensure it's called and throws errors when expected. Looks like you already added a test that ensures it's called, but still need tests for when errors are thrown.

> Is it a constraint for add2app that it should work without our generated xcconfig?

Oof, yes, add2app doesn't use a generated xcconfig. 

> Do a dry-run invocation inside KernelSnapshot to produce a native_assets.yaml there. 

So I'm a little confused. With add2app (Option A), there is no "outer" run. Apps are built and run through Xcode. There is only the "inner" assemble command. I'm I misunderstanding something?

Perhaps, in the "inner" assemble command, if `NATIVE_ASSETS` is not provided, it's does a dry run and non-dry-run if applicable? And then for non-add2app scenarios, we always set `NATIVE_ASSETS` in the xcconfig and if there are none, we set to the a value like `FALSE` or `NA` and skip if it's that value.

> Edit: A second issue seems to be that flutter clean removes the build folder

How does `flutter clean` come into play?

dcharkes commented:
> > Edit: A second issue seems to be that flutter clean removes the build folder
> 
> How does `flutter clean` come into play?

That is the location the `native_assets.yaml` was in. So it's both deleted and the path is not being passed through xcconfig.

> > Is it a constraint for add2app that it should work without our generated xcconfig?
> 
> Oof, yes, add2app doesn't use a generated xcconfig.

Gotcha, yet another requirement gathered. 🙊 

> > Do a dry-run invocation inside KernelSnapshot to produce a native_assets.yaml there.
> 
> So I'm a little confused. With add2app (Option A), there is no "outer" run. Apps are built and run through Xcode/Cocoapods. There is only the "inner" assemble command. Am I misunderstanding something?

Correct, we only have an "inner" assemble command.

> Perhaps, in the "inner" assemble command, if `NATIVE_ASSETS` is not provided, it's does a dry run and non-dry-run if applicable? And then for non-add2app scenarios, we always set `NATIVE_ASSETS` in the xcconfig and if there are none, we set to the a value like `FALSE` or `NA` and skip if it's that value.

The approach I'm working on right now is

>> 2. Make the KernelSnapshot target depend on the NativeAssets target and use the non-dry-run native_assets.yaml. The downside of this is that now KernelSnapshot caching doesn't work anymore. NativeAssets targets don't have caching.

This seems to work as intended almost all tests pass. As you can see in the latest failures on the CI there's one thing not working. Build caching does no longer work because the `NativeAssets` target hasn't implemented that.

I believe the right solution is to write out a deps file with the files that were read during compilation of native assets, so that the Flutter build system can skip the `NativeAssets` target if none of the files were changed. We already have tracking of which files were read during builds, but we need to make this list of files available to flutter_tools (https://github.com/dart-lang/native/issues/105).

dcharkes committed:
Merge branch 'main' into native-assets-macos-ios

dcharkes committed:
Run `flutter update-packages --force-upgrade`

dcharkes committed:
Fix analysis

dcharkes committed:
bump dependency to published version

dcharkes committed:
Merge branch 'main' into native-assets-macos-ios

dcharkes committed:
Run `flutter update-packages --force-upgrade`

dcharkes committed:
Undo `flutter update-packages --force-upgrade`

dcharkes committed:
Run regen hash

With
`flutter update-packages --cherry-pick-package=analyzer --cherry-pick-version=6.0.0`

dcharkes commented:
@stuartmorgan @vashworth 

The caching test seems to fail on the CI:

```
▌06:54:32▐ Testing macOS cache for dev/integration_tests/flutter_gallery...
RUNNING: cd dev/integration_tests/flutter_gallery; ../../../bin/flutter build macos --performance-measurement-file=perf.json --debug -v
[...]
ELAPSED TIME: 6.085s for ../../../bin/flutter build macos --performance-measurement-file=perf.json --debug -v in dev/integration_tests/flutter_gallery
╔═╡ERROR╞═══════════════════════════════════════════════════════════════════════
║ Not all build targets cached after second run.
║ The target performance data was: {"targets":[{"name":"gen_localizations","skipped":true,"succeeded":true,"elapsedMilliseconds":0},
║ {"name":"gen_dart_plugin_registrant","skipped":true,"succeeded":true,"elapsedMilliseconds":0},
║ {"name":"debug_macos_framework","skipped":true,"succeeded":true,"elapsedMilliseconds":0},
║ {"name":"debug_unpack_macos","skipped":true,"succeeded":true,"elapsedMilliseconds":263},
║ {"name":"native_assets","skipped":false,"succeeded":true,"elapsedMilliseconds":282},
║ {"name":"kernel_snapshot","skipped":false,"succeeded":true,"elapsedMilliseconds":1205},
║ {"name":"debug_macos_bundle_flutter_assets","skipped":true,"succeeded":true,"elapsedMilliseconds":658}]}
╚═══════════════════════════════════════════════════════════════════════════════
```

[full log](https://logs.chromium.org/logs/flutter/buildbucket/cr-buildbucket/8773325147804101297/+/u/run_test.dart_for_build_tests_shard_and_subshard_3_4/test_stdout)

However, locally these tests succeed:

```
$ cd /Users/dacoharkes/flt/engine/flutter  && SHARD=build_tests SUBSHARD=18_24 dart dev/bots/test.dart
▌11:01:05▐ STARTING ANALYSIS
▌11:01:05▐ SHARD=build_tests
▌11:01:05▐ SUBSHARD=18_24
▌11:01:05▐ Testing APK debug build for dev/integration_tests/flutter_gallery...
▌11:01:25▐ Testing APK cache for dev/integration_tests/flutter_gallery...
▌11:01:35▐ Testing APK release build for dev/integration_tests/flutter_gallery...
▌11:02:17▐ Testing APK cache for dev/integration_tests/flutter_gallery...
▌11:02:27▐ Testing IPA debug build for dev/integration_tests/flutter_gallery...
▌11:02:59▐ Testing IPA cache for dev/integration_tests/flutter_gallery...
▌11:03:07▐ Testing IPA release build for dev/integration_tests/flutter_gallery...
▌11:03:42▐ Testing IPA cache for dev/integration_tests/flutter_gallery...
▌11:03:49▐ Testing macOS debug build for dev/integration_tests/flutter_gallery...
▌11:04:15▐ Testing macOS cache for dev/integration_tests/flutter_gallery...
▌11:04:22▐ Testing macOS release build for dev/integration_tests/flutter_gallery...
▌11:05:06▐ Testing macOS cache for dev/integration_tests/flutter_gallery...
▌11:05:13▐ Test successful.
```

Any ideas or suggestions?

dcharkes committed:
Bump template dependencies

dcharkes committed:
Merge branch 'main' into native-assets-macos-ios

dcharkes committed:
Stop passing a `native_assets.yaml` path to flutter assemble

We build the native_assets.yaml inside the NativeAssets target inside flutter assemble.
This commit removes the dry-runs for the build command, as it solely relies on the backend.
This commit removes passing info through env vars and xcconfig.

dcharkes committed:
Fix template

dcharkes committed:
Add logging around `native_assets.yaml` existing

dcharkes committed:
Ensure `native_assets.yaml` exists in unit test

dcharkes committed:
Try figuring out why `native_assets.yaml` is not being created in `NativeAssets` target on CI

dcharkes committed:
Fix analysis

dcharkes committed:
Ensure writing a `native_assets.yaml` in MacOS if native assets is not enabled

dcharkes committed:
Test add2app bundles dylibs

dcharkes committed:
Call FFI in add2app

dcharkes committed:
Remove FFI call from add2app UI

dcharkes committed:
Merge commit 'bb6b72d6c66f62e2a13569241f4c7c711b5c5e9c' into native-assets-macos-ios

dcharkes committed:
Mock CCompilerConfig so tests can be randomized

dcharkes committed:
Unit test NativeAssets target

dcharkes committed:
Fix analysis

dcharkes committed:
Late initialize `cCompilerConfig` to avoid eagerly running

dcharkes commented:
Add2app is now also fully supported and tested. (Using the `native_assets.yaml` that is created inside `flutter assemble`.) This also removed a bunch of code for `flutter build` because it now doesn't have to know about native assets at all. Only the backend `flutter assemble`+wrapper scripts and `flutter run` (for hot restart support) need to know about native assets. @vashworth could you review the PR again?

dcharkes committed:
Merge branch 'main' into native-assets-macos-ios

dcharkes committed:
Fix imports after merge

christopherfujino commented on dev/devicelab/bin/tasks/module_test_ios.dart:
```diff
@@ -184,6 +199,11 @@ dependencies:
 
+      final File mainDart = File(path.join(projectDir.path, 'lib/main.dart'));
+      String mainDartContent = await mainDart.readAsString();
+      mainDartContent = mainDartContent.replaceAll('// UNCOMMENT: ', '');
```
TODO @christopherfujino to investigate ensuring that this does not leak state to subsequent tests

christopherfujino replied:
Ahh, we're doing a `flutter create` into a temp dir on lines 35-48, this is fine :)

eliasyishak commented:
friendly ping @christopherfujino @vashworth @stuartmorgan

stuartmorgan commented on packages/flutter_tools/lib/src/native_assets.dart:
```diff
@@ -0,0 +1,242 @@
+    );
+  }
+
+  /// Flutter expects `xcrun` to be on the path.
```
Why is something this macOS-specific in the main file?

stuartmorgan commented on packages/flutter_tools/lib/src/native_assets.dart:
```diff
@@ -0,0 +1,242 @@
+
+/// Mocks all logic instead of using `package:native_assets_builder`, which
+/// relies on doing process calls to `pub` and the local file system.
+class FakeNativeAssetsBuildRunner implements NativeAssetsBuildRunner {
```
Shouldn't this be in test code, rather than here?

stuartmorgan commented on packages/flutter_tools/lib/src/run_hot.dart:
```diff
@@ -440,6 +450,80 @@ class HotRunner extends ResidentRunner {
+    final FlutterDevice flutterDevice = flutterDevices.single;
+    final TargetPlatform targetPlatform = flutterDevice.targetPlatform!;
+
+    late Uri? nativeAssetsYaml;
```
`final` rather than `late`?

stuartmorgan commented on packages/flutter_tools/templates/package_ffi/lib/projectName.dart.tmpl:
```diff
@@ -0,0 +1,117 @@
+// To add platforms, run `flutter create -t plugin_ffi --platforms <platforms> .`
+// in this directory. You can also find a detailed instruction on how to
+// add platforms in the `pubspec.yaml` at
+// https://flutter.dev/docs/development/packages-and-plugins/developing-packages#plugin-platforms.
```
None of this applies here, does it?

stuartmorgan commented on packages/flutter_tools/templates/package_ffi/pubspec.yaml.tmpl:
```diff
@@ -0,0 +1,19 @@
+  sdk: {{dartSdkVersionBounds}}
+
+dependencies:
+  cli_config: ^0.1.1
```
Do we want to actively push a tool-level dependency into the package ecosystem? That seems like the wrong support level for something being used this way in our own template.

stuartmorgan replied:
> I believe we should graduate that package once we decide to take the native assets feature out of experimental.

If we haven't actually decided to take it out of experimental, then it really doesn't seem like we should be building templates for it into `flutter` that will encourage people to build and publish packages using it. If we don't actually have a decision, why not just publish a template as a one-off repository to point people to, instead of building the tooling into `flutter` first?

dcharkes replied:
We need support in `flutter run`/`flutter assemble`. I guess we could omit `flutter create` support.

Would it be possible to gate `flutter create` support behind the experimental flag as well?

stuartmorgan replied:
That seems reasonable.

devoncarew replied:
If the discussion here is about the criteria for graduating package:cli_config from tools.dart.dev to dart.dev, that's just gated on us believing the functionality is general purpose enough, and the current owner+team agreeing to maintaining the package at a high level SLO (https://pub.dev/publishers/dart.dev/packages vs https://pub.dev/publishers/tools.dart.dev/packages).

christopherfujino replied:
> We need support in `flutter run`/`flutter assemble`. I guess we could omit `flutter create` support.
> 
> Would it be possible to gate `flutter create` support behind the experimental flag as well?

This SGTM

dcharkes replied:
I have gated the template behind the feature.

> us believing the functionality is general purpose enough

I believe so. We wouldn't want every `build.dart` dev reinvent the wheel on:

1. How to parse the yaml config and output the yaml output in the communication with the Dart and Flutter SDK.
2. Reinvent the wheel on how to invoke a native C toolchain from a Dart script.

However, I believe package:native_toolchain_c can be made much more feature rich than it is now.

> the current owner+team agreeing to maintaining the package at a high level SLO

Packages: cli_config and native_assets_cli we (the native interop team) will maintain with high SLO. I expect most contributions to come from ourselves. These packages are the counterpart of this PR and the CLs in the Dart SDK in the protocol between the SDKs and the packages.

For package native_toolchain_c, I expect that to be more on the reviewers side as users start demanding more features from it and submit PRs. We will shepherd these contributions.

stuartmorgan commented on packages/flutter_tools/test/general.shard/build_system/targets/common_test.dart:
```diff
@@ -75,6 +75,15 @@ void main() {
+  - 0
+  - 0
+native-assets: {}
+''');
```
Looks like this would benefit from a helper so we're not copy/pasting this a bunch of times.

dcharkes committed:
Merge branch 'main' into native-assets-macos-ios

dcharkes committed:
address comments

dcharkes committed:
Move FakeNativeAssetsBuildRunner to test dir

dcharkes committed:
Move cCompilerConfig for macOS to macOS file

dcharkes committed:
fix xcode_backend.dart

dcharkes committed:
address comments

dcharkes committed:
rename var

dcharkes committed:
Move shared logic out of macos/native_assets.dart

dcharkes committed:
split generate FFI plugin and package

dcharkes committed:
address comment

dcharkes committed:
Remove platforms argument from create commands

dcharkes committed:
Stop generating os-specific folders

dcharkes reviewed (commented):
Thanks @stuartmorgan!

I've addressed most comments, see my replies to the remaining comments.

stuartmorgan commented on dev/devicelab/bin/tasks/module_test_ios.dart:
```diff
@@ -160,6 +160,19 @@ Future<void> main() async {
       await _createFakeDartPlugin(dartPluginName, tempDir);
 
+      // Make a package with native assets.
+      await inDirectory(projectDir, () async {
```
It sounds like you are describing the reason for calling `config`; my comment is about the `inDirectory` wrapper.

vashworth commented on packages/flutter_tools/test/general.shard/ios/native_assets_test.dart:
```diff
@@ -0,0 +1,242 @@
+          Package('bar', projectUri),
+        ],
+      ),
+    );
```
You don't have any expectations for this test? Wouldn't you expect the returned list of Uris to be empty, a yaml file created in the file system, no assets created, etc?

dcharkes committed:
Remove inDirectory

dcharkes committed:
Fix formatting

stuartmorgan commented on dev/integration_tests/ios_host_app/flutterapp/lib/main:
```diff
@@ -7,6 +7,7 @@ import 'dart:ui' as ui;
 
 import 'package:flutter/material.dart';
 import 'package:flutter/services.dart';
+// UNCOMMENT: import 'package:ffi_package/ffi_package.dart';
```
> Because the test harness checks that add-to-app works by clicking buttons, so all that logic in the main file needs to be there.

We wouldn't need to re-run those same tests a second time though. I'm saying we can build a project, run the existing tests, replace main, and then run different, FFI-specific tests. Am I missing something about the flow that makes that infeasible?

dcharkes replied:
There are two sections before all the plugins are started:

```dart
section('Build ephemeral host app in release mode without CocoaPods');

// ... run a `flutter build`

// Check the tool is no longer copying to the legacy xcframework location.
checkDirectoryNotExists(path.join(projectDir.path, '.ios', 'Flutter', 'engine', 'Flutter.xcframework'));
```

And some extra checks that the build succeeded.

Maybe we can just move those checks until after the extra plugins/packages are added. I'll take a look.

dcharkes replied:
I have moved the checks until after the native assets package and other plugins are added. The `// UNCOMMENT: ` logic has been removed.

vashworth commented on packages/flutter_tools/test/general.shard/ios/native_assets_test.dart:
```diff
@@ -0,0 +1,242 @@
+        environment.projectDir.childFile('.dart_tool/package_config.json');
+    await packageConfig.parent.create();
+    await packageConfig.create();
+    await buildNativeAssetsiOS(
```
Same with this one, no expectations?

vashworth commented on packages/flutter_tools/test/general.shard/macos/native_assets_test.dart:
```diff
@@ -0,0 +1,318 @@
+    );
+  });
+
+  testUsingContext('dry run with no package config',
```
This test has the same name as the first one in this file, perhaps name it something like "dry run for multiple OSes with no package config"

vashworth commented on packages/flutter_tools/test/general.shard/macos/native_assets_test.dart:
```diff
@@ -0,0 +1,318 @@
+      ),
+    );
+  });
+}
```
You should add a test to ensure `ensureNoLinkModeStatic` is being used as expected.

vashworth reviewed (commented):
Another friendly reminder to consult the style guide: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo

Couple things I noticed commonly:
* `=>` [should only be used when it can fit on one line](https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#consider-using--for-short-functions-and-methods)
* [Indent multi-line argument and parameter lists by 2 characters](https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#indent-multi-line-argument-and-parameter-lists-by-2-characters)
* [Prefer avoiding line breaks after assignment operators.](https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#prefer-a-maximum-line-length-of-80-characters)
* [Use a trailing comma for arguments, parameters, and list items, but only if they each have their own line... [if] everything would fit on one line, put it all on one line, otherwise, split it one item to a line.](https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#use-a-trailing-comma-for-arguments-parameters-and-list-items-but-only-if-they-each-have-their-own-line)

vashworth commented on .ci.yaml:
```diff
@@ -3978,13 +3978,32 @@ targets:
+      tags: >
+        ["devicelab", "ios", "mac"]
+      task_name: native_assets_ios
+
```
Oh wait, you do test it in the integration shard, right? My bad, I added this at the beginning of my review

vashworth replied:
Although, looking at the README for integration.shard testing, I think you're only supposed to run against `flutter_tester` in the integration.shard

https://github.com/vashworth/flutter/blob/master/packages/flutter_tools/test/integration.shard/README.md

So maybe it would be better to make a macOS integration test

vashworth replied:
@christopherfujino Do you have an opinion on this?

christopherfujino replied:
> @christopherfujino Do you have an opinion on this?

It looks like this *is* using flutter-tester? https://github.com/flutter/flutter/pull/130494/files#diff-23a9696abb2839217a91641a47e161c2e7a3010989874a14f2ba0b78221d5d0fR28 Or were you referring to a different test?

vashworth replied:
I can't tell what that link points to (maybe the PR is too big or something).

I was referring to this test, which tests both `flutter-tester` and the host device
https://github.com/flutter/flutter/blob/fae782f5a588cb08230ce082c8c67898b49defab/packages/flutter_tools/test/integration.shard/native_assets_test.dart#L27-L30

christopherfujino replied:
testing a desktop app is not an issue, this LGTM (if I had to guess, it would be that the README in question is saying that you need the devicelab to run a test with a phone).

dcharkes commented on packages/flutter_tools/lib/src/commands/create.dart:
```diff
@@ -385,7 +388,16 @@ class CreateCommand extends CreateBase {
+          projectType: template,
+        );
+        pubContext = PubContext.createPlugin;
+      case FlutterProjectType.packageFfi:
```
I unshared the code between `_generateFfiPluginOrPackage` because Stuart pointed out a lot is not shared.

dcharkes committed:
Gate test behind isMacOS

dcharkes committed:
address comments

dcharkes committed:
expect process calls in test

dcharkes committed:
Expect no process invocations in test

dcharkes committed:
Static libraries error test

dcharkes committed:
Add test expectations

dcharkes committed:
hot runner test exception

dcharkes committed:
Skip tests with mac process invocations on windows

dcharkes committed:
Gate flutter create --template=package_ffi behind experimental flag

dcharkes committed:
Remove `// UNCOMMENT: ` stuff from module test

dcharkes committed:
Don't cache allowed templates for create command

dcharkes committed:
Formatting assignment no newlines

dcharkes committed:
Hand format `=>`s

dcharkes committed:
`MacOS` and `iOS` spelling in method names

dcharkes committed:
Lint on same line

dcharkes committed:
Fix formatting

dcharkes committed:
Merge branch 'main' into native-assets-macos-ios

dcharkes committed:
Remove if

dcharkes committed:
Revert "Remove if"

This reverts commit 80c23af180142827a4320cbc3d435d7552f87993.

dcharkes committed:
Add comment

dcharkes committed:
Merge branch 'main' into native-assets-macos-ios

vashworth commented on packages/flutter_tools/lib/src/ios/native_assets.dart:
```diff
@@ -0,0 +1,153 @@
+  }
+
+  final Uri buildUri_ = nativeAssetsBuildUri(projectUri, OS.iOS);
+  final Iterable<Asset> assetTargetLocations = await dryRunNativeAssetsIOSInternal(fileSystem, projectUri, buildRunner);
```
Formatting: [Use a trailing comma for arguments, parameters, and list items, but only if they each have their own line... [if] everything would fit on one line, put it all on one line, otherwise, split it one item to a line.](https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#use-a-trailing-comma-for-arguments-parameters-and-list-items-but-only-if-they-each-have-their-own-line)

So for example, since this line doesn't all fit on one line within 80 characters, use a trailing command and give each item its own line.

```dart
final Iterable<Asset> assetTargetLocations = await dryRunNativeAssetsIOSInternal(
  fileSystem, 
  projectUri, 
  buildRunner,
);
```

There's a couple of other places you can do this in this file

vashworth commented on packages/flutter_tools/lib/src/macos/native_assets_host.dart:
```diff
@@ -0,0 +1,141 @@
+    throwToolExit('Failed to find clang with xcrun:\n${xcrunResult.stderr}');
+  }
+  final String installPath = (xcrunResult.stdout as String)
+    .split('\n')
```
Use `LineSplitter.split` to split on line break

vashworth commented on packages/flutter_tools/lib/src/native_assets.dart:
```diff
@@ -0,0 +1,298 @@
+
+  @override
+  Future<bool> hasPackageConfig() {
+    final File packageConfigJson = fileSystem.directory(projectUri.toFilePath()).childFile('.dart_tool/package_config.json');
```
```dart
.childDirectory('.dart_tool')
.childFile('package_config.json');
```

dcharkes committed:
Address comments

dcharkes commented on dev/devicelab/bin/tasks/module_test_ios.dart:
```diff
@@ -489,10 +519,6 @@ end
-            'CODE_SIGNING_ALLOWED=NO',
-            'CODE_SIGNING_REQUIRED=NO',
-            'CODE_SIGN_IDENTITY=-',
-            'EXPANDED_CODE_SIGN_IDENTITY=-',
```
The tests seem to pass on the CI without these lines. And they enable running the tests locally for me. We could consider just removing them. (Added them back now.)

dcharkes committed:
Merge branch 'main' into native-assets-macos-ios

vashworth commented on packages/flutter_tools/test/general.shard/flutter_project_metadata_test.dart:
```diff
@@ -184,4 +187,14 @@ migration:
+    expect(FlutterProjectType.enabledValues, contains(FlutterProjectType.packageFfi));
+  }, overrides: <Type, Generator>{
+    FeatureFlags: () => TestFeatureFlags(isNativeAssetsEnabled: true),
+  });
```
Can you also test a non-native assets feature flag to ensure it's added properly?

vashworth reviewed (commented):
Have you looked into how to fix with G3 yet?

dcharkes commented:
> Have you looked into how to fix with G3 yet?

@iinozemtsev was so kind to make one! cl/563109654
I've added it in the FROB UI, but I didn't see anything happening. Should it show up in the UI?

dcharkes committed:
Address comment

vashworth commented:
> > Have you looked into how to fix with G3 yet?
> 
> @iinozemtsev was so kind to make one! cl/563109654 I've added it in the FROB UI, but I didn't see anything happening. Should it show up in the UI?

I see it in the FROB UI now, let's see if it passes this time

vashworth reviewed (approved):
Sorry for all the delays, LGTM!

dcharkes commented:
> > > Have you looked into how to fix with G3 yet?
> > 
> > 
> > @iinozemtsev was so kind to make one! cl/563109654 I've added it in the FROB UI, but I didn't see anything happening. Should it show up in the UI?
> 
> I see it in the FROB UI now, let's see if it passes this time

```
UnusedValueUtils.kt:433:42: error: cannot find a parameter with this name: allowErrorTypes
        samType.asPsiType(psiTypeParent, allowErrorTypes = true) as? PsiClassType
```

The g3 error seems unrelated.

dcharkes committed:
Merge branch 'main' into native-assets-macos-ios

christopherfujino reviewed (approved):
LGTM

dcharkes commented:
@vashworth Should the google testing take 2 hours to run?

vashworth commented:
> @vashworth Should the google testing take 2 hours to run?

In FRoB, it says "Rebase required" - I've never seen that before but maybe try rebasing?

christopherfujino committed:
Merge branch 'master' into native-assets-macos-ios

dcharkes commented:
I would usually think that merging is not rebasing. I can try to just squash all commits, rebase on master as single commit and force push to "rebase". (But I believe force pushing to PRs under review is frowned upon as the diff is no longer visible.)

dcharkes commented:
> G3Fix failed to apply

cl/563109654 probably needs to be rebased as well.

christopherfujino commented:
> I would usually think that merging is not rebasing. I can try to just squash all commits, rebase on master as single commit and force push to "rebase". (But I believe force pushing to PRs under review is frowned upon as the diff is no longer visible.)

You're right they're different, but in terms of FRoB, the difference doesn't matter.

dcharkes committed:
Merge branch 'main' into native-assets-macos-ios

dcharkes commented:
Woohoo, all green!

@stuartmorgan waiting on approval from your side.

dcharkes committed:
Merge branch 'main' into native-assets-macos-ios

zanderso commented on packages/flutter_tools/lib/src/run_hot.dart:
```diff
@@ -21,6 +21,9 @@ import 'dart/package_map.dart';
 import 'devfs.dart';
 import 'device.dart';
 import 'globals.dart' as globals;
+import 'ios/native_assets.dart';
```
I'm a bit confused by how the PRs on this are organized, but I left a comment on another similar PR about the layering violation here.

dcharkes replied:
I don't believe GitHub has a solution for stacked pull requests. So the other PRs based on top of this one show the whole diff including this one. If there's a better workflow on GitHub I'd love to learn about it.

I've looked around what to do about the layering violation here.
It doesn't really make sense to push native assets logic into the device or even the resident compiler. The resident compiler doesn't know about the target. It's a property of the hot-runner that it has a second kernel compiler (if you count the one in the flutter assemble backend as first kernel compiler). So it should be responsible for embedding the native_assets.yaml in the kernel file.
I have moved the branching based on device to `src/native_assets.dart` now, so that it includes the various `src/<os>/native_assets.dart`.

dcharkes committed:
Merge branch 'main' into native-assets-macos-ios

dcharkes committed:
Move OS branching to native_assets.dart to run_hot.dart

stuartmorgan-g reviewed (approved):
LGTM. I didn't do a complete re-review since people with more domain expertise have already approved, but the changes based on my previous comments look good!

dcharkes committed:
Merge branch 'main' into native-assets-macos-ios
